File: server/src/routes/user/address/schema.rs
use serde::{Deserialize, Serialize};
use utoipa::{ToResponse, ToSchema};
use validator::Validate;

#[derive(Debug, Deserialize, Serialize, ToSchema, ToResponse, Validate)]
#[serde(rename_all = "camelCase")]
pub struct CreateAddressInput {
    #[validate(
        length(min = 1, message = "First name is required"),
        length(max = 20, message = "Maximum 20 characters are allowed")
    )]
    #[schema(example = "John", min_length = 1, max_length = 20)]
    /// First name of the user.
    pub first_name: String,

    #[validate(
        length(min = 1, message = "First name is required"),
        length(max = 20, message = "Maximum 20 characters are allowed")
    )]
    #[schema(example = "Doe", min_length = 1, max_length = 20)]
    /// Last name of the user.
    pub last_name: String,

    #[validate(
        length(min = 1, message = "Phone number is required"),
        length(equal = 10, message = "Phone number must be 10 digits")
    )]
    #[schema(example = "1234567890", min_length = 1, max_length = 10)]
    /// Phone number of the user.
    pub phone_number: String,

    #[validate(
        length(min = 1, message = "Full Address is required"),
        length(min = 5, message = "Full Address must be at least 5 characters long"),
        length(max = 300, message = "Maximum 300 characters are allowed")
    )]
    #[schema(min_length = 1, max_length = 300)]
    /// Address of the user.
    pub full_address: String,

    #[validate(
        length(min = 1, message = "City is required"),
        length(max = 50, message = "Maximum 50 characters are allowed")
    )]
    #[schema(min_length = 1, max_length = 50)]
    /// City of the user.
    pub city: String,

    #[validate(
        length(min = 1, message = "State is required"),
        length(max = 50, message = "Maximum 50 characters are allowed")
    )]
    #[schema(min_length = 1, max_length = 50)]
    /// State of the user.
    pub state: String,

    #[validate(range(min = 100000, max = 999999, message = "Postal code is invalid"))]
    #[schema(example = 123456, minimum = 100000, maximum = 999999)]
    /// Postal Code of the user.
    pub postal_code: i32,

    #[validate(
        length(min = 1, message = "Landmark is required"),
        length(max = 50, message = "Maximum 50 characters are allowed")
    )]
    #[schema(min_length = 1, max_length = 50)]
    /// Landmark of the user.
    pub landmark: Option<String>,
}

File: client/apps/www/src/app/(main)/user/addresses/_lib/schema.ts
import { type CreateAddressInput } from "@/client";
import { coerceNumber, string, withSchema } from "@repo/utils/zod";
import { literal, number } from "zod";

export const createAddressSchema = withSchema<CreateAddressInput>()({
  firstName: string()
    .min(1, "First Name is required")
    .max(20, "Maximum 20 characters are allowed"),
  lastName: string()
    .min(1, "Last Name is required")
    .max(20, "Maximum 20 characters are allowed"),
  city: string()
    .min(1, "City is required")
    .max(50, "Maximum 50 characters are allowed"),
  fullAddress: string()
    .min(1, "Full Address is required")
    .max(300, "Maximum 300 characters are allowed"),
  landmark: string()
    .min(1, "Landmark is required")
    .max(200, "Maximum 200 characters are allowed")
    .or(literal("")),
  phoneNumber: string().min(1, "Phone Number is required"),
  postalCode: coerceNumber(number(), "Postal Code is required"),
  state: string()
    .min(1, "State is required")
    .max(50, "Maximum 50 characters are allowed"),
});

createAddressSchema._type satisfies CreateAddressInput;
